File: python/arrays/ProductOfArrayExceptSelf.py
'''
Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].
The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
You must write an algorithm that runs in O(n) time and without using the division operation.

Input: nums = [1,2,3,4]
Output: [24,12,8,6]

Input: nums = [-1,1,0,-3,3]
Output: [0,0,9,0,0]
'''
from typing import List

class ProductOfArrayExceptSelf:
    """
    Time Complexity: O(N) - Two passes over the array.
    Space Complexity: O(1) - Excluding the output array.
    """
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [1] * n
        
        # 1. Calculate Prefix Products
        # res[i] will contain product of nums[0]...nums[i-1]
        prefix = 1
        for i in range(n):
            res[i] = prefix
            prefix *= nums[i]
            
        # 2. Calculate Suffix Products and multiply with Prefix
        # postfix will contain product of nums[i+1]...nums[n-1]
        postfix = 1
        for i in range(n - 1, -1, -1):
            res[i] *= postfix
            postfix *= nums[i]
            
        return res

if __name__ == "__main__":
    solution = ProductOfArrayExceptSelf()
    
    # Test Case 1
    nums1 = [1, 2, 3, 4]
    print(f"Input: {nums1}")
    print(f"Output: {solution.productExceptSelf(nums1)}") # Expected: [24, 12, 8, 6]
    print("-" * 20)

    # Test Case 2
    nums2 = [-1, 1, 0, -3, 3]
    print(f"Input: {nums2}")
    print(f"Output: {solution.productExceptSelf(nums2)}") # Expected: [0, 0, 9, 0, 0]

File: src/main/java/com/naresh/a_dsalgo/aa_arrays/problems/ProductOfArrayExceptSelf.java
package com.naresh.a_dsalgo.aa_arrays.problems;

import java.util.Arrays;

/**
 * Problem: Product of Array Except Self
 * Description: Return an array where each element is the product of all elements except nums[i].
 */
public class ProductOfArrayExceptSelf {
    /**
     * Algorithm: Two-pass accumulation using preFix and postFix running products.
     */
    public int[] productExceptSelf(int[] nums) {
        // Pattern: Prefix & Postfix Product | Time: O(n), Space: O(1)
        var n = nums.length;
        var res = new int[n];
        var preFix = 1;
        for (var i = 0; i < n; i++) {
            res[i] = preFix; // Store product of all left elements
            preFix *= nums[i];
        }
        var postFix = 1;
        for (var i = n - 1; i >= 0; i--) {
            res[i] *= postFix; // Multiply by product of all right elements
            postFix *= nums[i];
        }
        return res;
    }
    // FAANG Tip: Avoid division to handle zeros. Output array doesn't count towards space complexity.

    public static void main(String[] args) {
        var solution = new ProductOfArrayExceptSelf();
        int[] nums = {1, 2, 3, 4};
        System.out.println("Output: " + Arrays.toString(solution.productExceptSelf(nums)));
    }
}

/**
 * Dry Run:
 * Input: nums = {1, 2, 3, 4}
 *
 * 1. Initialization:
 *    n = 4, res = [0, 0, 0, 0], preFix = 1
 *
 * 2. First Pass (Prefix Products):
 *    i = 0: res[0] = 1, preFix = 1
 *    i = 1: res[1] = 1, preFix = 2
 *    i = 2: res[2] = 2, preFix = 6
 *    i = 3: res[3] = 6, preFix = 24
 *    Result: [1, 1, 2, 6]
 *
 * 3. Second Pass (Postfix Products):
 *    postFix = 1
 *    i = 3: res[3] = 6 * 1 = 6, postFix = 4
 *    i = 2: res[2] = 2 * 4 = 8, postFix = 12
 *    i = 1: res[1] = 1 * 12 = 12, postFix = 24
 *    i = 0: res[0] = 1 * 24 = 24, postFix = 24
 *    Result: [24, 12, 8, 6]
 */
